File: lab_11/lab_09_2_before.R
# Esli russkie bukvi prevratilitis v krakozyabry, to File - Reopen with
# encoding... - UTF-8 - Set as default - OK

# lab 09-2

# подключаем пакеты
library(caret) # стандартизованный подход к регрессионным и классификационным моделям
library(AER) # инструментальные переменные
library(sandwich) # робастные стандартные ошибки
library(ivpack) # дополнительные плющки для инструментальных переменных
# library(memisc) # табличка mtable
library(rio) # импорт файлов разных форматов
library(tidyverse) # графики и манипуляции с данными, подключаются пакеты dplyr, ggplot2, etc
library(huxtable) # сравнение нескольких табличек
library(estimatr) # robust se

# оценивание заданной формы модели с эндогенностью
data('CigarettesSW', package = 'AER')
h = CigarettesSW
glimpse(h)
help('CigarettesSW')

qplot(data = h, x = price, y = packs)
h95 = filter(h, year == '1995')

h2 = mutate(h95,
            rprice = price / cpi,
            rincome = income / (cpi * population),
            tdiff = (taxs - tax) / cpi)
qplot(data = h2, x = price, y = packs)
qplot(data = h2, x = rprice, y = packs)

# МНК
model_0 = lm(data = h2, log(packs) ~ log(rprice))
summary(model_0)


# 2МНК руками:
stage_1 = lm(data = h2, log(rprice) ~ tdiff)
stage_1
h3 = mutate(h2, log_price_hat = fitted(stage_1))
stage_2 = lm(data = h3, log(packs) ~ log_price_hat)

stage_2
summary(stage_2) # se лгут!

model_iv = ivreg(data = h2,
      log(packs) ~ log(rprice) | tdiff)
summary(model_iv)

model_iv_hc = iv_robust(data = h2,
                 log(packs) ~ log(rprice) | tdiff)
summary(model_iv_hc)


huxreg(model_0, model_iv, model_iv_hc)


model_iv_hc2 = iv_robust(data = h2,
      log(packs) ~ log(rprice) + log(rincome)
      | tdiff + log(rincome))
summary(model_iv_hc2)

File: lab_11/lab_09_2_before.py
#!/usr/bin/env python
# coding: utf-8

# In[42]:


import numpy as np # математические операции
import pandas as pd # операции с табличками
import statsmodels.api as sm # модели 
import statsmodels.formula.api as smf # модели 
from sklearn.linear_model import LinearRegression # линейная регрессия без статистики
import seaborn as sns # графики
import matplotlib.pyplot as plt # графики
from statsmodels.iolib.summary2 import summary_col # несколько моделей в одной таблице
from statsmodels.graphics.mosaicplot import mosaic


# In[ ]:


get_ipython().system('pip install rdatasets # наборы данных')
# !pip install yfinance
from rdatasets import data 

get_ipython().system('pip install linearmodels # IV')
from linearmodels.iv import IV2SLS


# In[43]:


cig = pd.read_csv('https://github.com/vincentarelbundock/Rdatasets/raw/master/csv/AER/CigarettesSW.csv')
cig.head()


# In[44]:


cig['rprice'] = cig['price'] / cig['cpi']
cig['rincome'] = cig['income'] / cig['cpi'] / cig['population']
cig['tdiff'] = cig['taxs'] - cig['tax']
cig['log_rprice'] = np.log(cig['rprice'])
cig['log_packs'] = np.log(cig['packs'])


# In[45]:


c95 = cig[cig['year'] == 1995]
c95.head()


# In[46]:


sns.relplot(data=c95, x='rprice', y='packs');


# In[47]:


# наивная модель
naive = smf.ols('log_packs ~ log_rprice', c95).fit()
naive.summary()


# In[48]:


# 2OLS руками
stage_1 = smf.ols('log_rprice ~ tdiff', c95).fit()
stage_1.summary()


# In[49]:


c95['log_rprice_hat'] = stage_1.fittedvalues


# In[50]:


c95.head()


# In[51]:


stage_2 = smf.ols('log_packs ~ log_rprice_hat', c95).fit()
stage_2.summary()


# In[60]:


# dependent ~ exog regressors + [endog regressors ~ instruments]
# указываем 1
formula = 'log_packs ~ 1 + [log_rprice ~ tdiff]'
iv_a = IV2SLS.from_formula(formula, c95).fit(cov_type='unadjusted')
iv_a


# In[61]:


iv_b = IV2SLS.from_formula(formula, c95).fit(cov_type='robust')
iv_b


# In[ ]:


formula = 'log_packs ~ 1 + rincome + [log_rprice ~ tdiff]'

